@assert

StefanHenke opened the pull request:
Describes the basic feature set of the new `@assert` annotation. The full blown CXN documentation is pending and will be provided by Patrice later. Once available, we will link it here. That´s why the expressions in the examples are kept simple by intention.

StefanHenke commented:
> Please introduce in smaller steps
> 
> * one case
> * multiple cases
> * error function
> 
> cosider using `? :`
> 
> Please add some examples for moderately more complex expressions.
> 
> Like, e.g.
> 
> * "if an order has items the order's total amount must be larger then 0"
> * "a book's author's name must not be null"

What exactly do you mean with `consider using `? :`?

We agreed with Johannes and Patrice yesterday that we will keep the expressions simple here and wait for the CXL docs. But I can check for a "moderately more complex expression".

rjayasinghe commented:
> > Please introduce in smaller steps
> > 
> > * one case
> > * multiple cases
> > * error function
> > 
> > cosider using `? :`
> > Please add some examples for moderately more complex expressions.
> > Like, e.g.
> > 
> > * "if an order has items the order's total amount must be larger then 0"
> > * "a book's author's name must not be null"
> 
> What exactly do you mean with `consider using `? :`?
> 
> We agreed with Johannes and Patrice yesterday that we will keep the expressions simple here and wait for the CXL docs. But I can check for a "moderately more complex expression".

I think @agoerler meant that we split up this single block into 3 examples. After each example we explain the nature of the used expression.

StefanHenke commented:
> > > Please introduce in smaller steps
> > > 
> > > * one case
> > > * multiple cases
> > > * error function
> > > 
> > > cosider using `? :`
> > > Please add some examples for moderately more complex expressions.
> > > Like, e.g.
> > > 
> > > * "if an order has items the order's total amount must be larger then 0"
> > > * "a book's author's name must not be null"
> > 
> > 
> > What exactly do you mean with `consider using `? :`?
> > We agreed with Johannes and Patrice yesterday that we will keep the expressions simple here and wait for the CXL docs. But I can check for a "moderately more complex expression".
> 
> I think @agoerler meant that we split up this single block into 3 examples. After each example we explain the nature of the used expression.

Yes, I have split up the examples already. But was not sure if `?:` is something in addition

patricebender commented on guides/providing-services.md:
```diff
@@ -939,6 +939,60 @@ Cross-service checks are not supported. It is expected that the associated entit
+Refer to [Expressions as Annotation Values](../cds/cdl.md#expressions-as-annotation-values) for detailed rules on expression syntax.
+
+::: info Expression Evaluation
+Expressions are evaluated *after* the request has been applied to the underlying datastore. The affected entities are the entities being part of the request and identified by their primary keys. The evaluating runtime constructs and executes statements with the annotation-provided expressions and the respective primary key values for the given entities. Depending on the structure and cardinality  ...
```
```suggestion
Expressions are evaluated *after* the request has been applied to the underlying datastore. The affected entities are the entities being part of the request and identified by their primary keys.
The evaluating runtime constructs and executes statements with the annotation-provided expressions and the respective primary key values for the given entities. Depending on the structure and cardinality of the request data it can lead to one or more statements being executed.
```

patricebender reviewed (commented):
👍 As Adrian stated, it would be nice to have a little more (complex) real life scenarios. I would have probably taken the well known Bookshop for showcasing the different scenarios. But that may be my personal taste.

StefanHenke commented:
> 👍 As Adrian stated, it would be nice to have a little more (complex) real life scenarios. I would have probably taken the well known Bookshop for showcasing the different scenarios. But that may be my personal taste.

I have improved the examples to use Bookshop related entities. Regarding the ternary operator, we still have to put the `: else` as the compiler complains otherwise.

agoerler commented on guides/providing-services.md:
```diff
@@ -939,6 +939,86 @@ Cross-service checks are not supported. It is expected that the associated entit
 
+### `@assert` <Beta/>
+
+Annotate an element with `@assert` to define an expression for more complex validations that need to be fulfilled before data gets persisted. If the validation should fail, the expression must return the error message to be sent to the client, or `null` if the validation passed.
```
> fulfilled before data gets persisted

I don't like this wording as we check after the data is persisted

* "fulfilled after the data has been persisted" - this is confusing if you consider the commit as part of the insert/update processing; still I like this wording best
* "fulfilled before the transaction is committed" - this is confusing as we don't validate before-commit (but we might adapt the implementation, to do so).
* "fulfilled eventually" - this is too fuzzy and could make readers believe we only support eventual consistency

agoerler commented on guides/providing-services.md:
```diff
@@ -939,6 +939,86 @@ Cross-service checks are not supported. It is expected that the associated entit
+    parent    : Association to Orders;  
+    book      : Association to Books;
+    @assert: (case
+      when book.stock <= amount then 'Stock exceeded'
```
> 'Stock exceeded'

I think we gave the recommendation that error messages should not contain element names and should be rather imperative like "order less books". See https://github.wdf.sap.corp/cds-java/home/issues/1164

But I am not sure wether this recommendation should also apply to `@ssert` messages.

@beckermarc - what do you think?

agoerler commented on guides/providing-services.md:
```diff
@@ -939,6 +939,86 @@ Cross-service checks are not supported. It is expected that the associated entit
+    end)
+    quantity  : Integer;
+    amount    : Decimal(9, 2);
+}
```
Maybe use manged compositions of many aspects:

```suggestion
entity Books : cuid {
   title : String;
   stock : Integer;
}

entity Orders : cuid {
   Items : Composition of many {
      key itemId   : GUID;
         book     : Association to Books;

         @assert: (case // [!code focus]
            when book.stock <= quantity then 'Stock exceeded' // [!code focus]
         end) // [!code focus]
         quantity : Integer;
         amount   : Decimal(9, 2);
   };
}
```

sjvans commented on guides/providing-services.md:
````diff
@@ -939,6 +939,86 @@ Cross-service checks are not supported. It is expected that the associated entit
+    @assert: ((book.stock <= quantity) ? 'Stock exceeded' : null)
+    quantity  : Integer;
+}
+```
````
however, the collector isn't described here and also not on by default, so i guess it's fine. we want to start out with "do it on service level" anyways, right?

agoerler reviewed (commented):
I think we should mention that `@assert` should be used on service level.

StefanHenke commented on guides/providing-services.md:
```diff
@@ -939,6 +939,110 @@ Cross-service checks are not supported. It is expected that the associated entit
+entity OrderItems : cuid {
+        
+  @assert: (case 
+    when quantity <= 0 then 'Quantity must be greater then zero' 
```
```suggestion
    when quantity <= 0 then 'Quantity must be greater than zero' 
```

Akatuoro commented on guides/providing-services.md:
```diff
@@ -939,6 +939,135 @@ Cross-service checks are not supported. It is expected that the associated entit
+  @assert: (case
+    when book.stock = 0 then 'Stock is zero'
+    when book.stock <= quantity then 'Stock exceeded'
+  end)
```
This assert should also be on service level since it contains second element / association
